designs/auto-certs-renewal.md

```diff
@@ -0,0 +1,122 @@
+
+### Worker node certificates
+
+1. just ca.crt in /etc/kubernetes/pki, which should be generated after `kubeadm join`
+
+## Goal
+
+1. Auto renew all the above mentioned certificates except etcd CA and kube CA, which last for 10 years.
+
+### Observation
+
+1. We can renew all etcd’s certs by simply **recreating the external etcd machine**. If doing manually, we just need to delete the auto generated certs and run `sudo etcdadm join phase certificates`
+2. We can renew all cp’s certs besides apiserver-etcd-client by **recreating the cp machine.** As we can see from the following diagram that the apiserver-etcd-client is generated by Etcdadm Controller, so this cert cannot be renewed by just recreating the cp machine.
+3. No need to do auto renew certs in worker nodes.
+
+## Solution
+
+Etcd can be deployed in [two topologies.](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/) Certs renew mechanism is slightly different in those two scenarios.
+
+### Stacked etcd Nodes
+
+In Stacked etcd nodes topology, etcd is running inside the cp node. 
+
+We can renew all certificates by just redeploying cp nodes
+
+1. Expose [rolloutBefore.certificatesExpiryDays](https://cluster-api.sigs.k8s.io/tasks/certs/auto-rotate-certificates-in-kcp.html) field from KubeadmControlPlane in EKSA cluster spec. Add pass the field to KubeadmControlPlane
+
+### External etcd Nodes
+
+In external etcd nodes topology, etcd is running in a separate machine besides cp machine. In addition to roll out cp nodes, we also need to roll out etcd nodes, with the following changes on top of above:
+
+1. Add rolloutBefore.certificatesExpiryDays field to [EtcdadmCluster](https://github.com/aws/etcdadm-controller/blob/main/api/v1beta1/etcdadmcluster_types.go#L38), exposing and passing it from EKSA. The business logic of this field and reconciler should be:
+    1. apply machine annotation `machine.cluster.x-k8s.io/certificates-expiry` during etcd node creation.
+    2. check if roll out is needed. Refer https://github.com/kubernetes-sigs/cluster-api/blob/main/controlplane/kubeadm/internal/controllers/controller.go#L380.
+    3. if roll out is needed for any etcd node, upgrade all etcd nodes.
+
+However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. We need to fix this bug by changing [etcdadm controller’s code](https://github.com/aws/etcdadm-controller/blob/main/controllers/certs.go#L81), it should update apiserver-etcd-client secret constantly in a short interval, such as every 24 hours. If we choose just update the secret during etcd nodes rolling out, then we have to invent mechanism to make sure the etcd nodes rolling out happens before cp nodes rolling out when it comes to renew certificates.
```

Comment:
> However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. 

Could you clarify this. Are we making a statement about the existing behavior or something that's the result of implementing part of a solution?

In conjunction with the clarification, I think we need to clearly state the problem before mentioning any solutions. Does the final sentence elaborate further on the problem for example? Perhaps combine that with the problem description from the first sentence.